File: store.py
import mongocache
import pymongo
import random
import os

import helpers

games = pymongo.MongoClient(os.getenv('MONGODB_URI')).get_database().games
games_cache = mongocache.CollectionCache(games, 'id')

def create_game(game_id, player_id, player_name):
  games.insert_one({
    'id': game_id,
    'creator': player_name,
    'players': [helpers.create_player(player_id, player_name)],
    'quests': []
  })
  return True

def join_game(game_id, player_id, player_name):
  result = games.update_one(
    {'id': game_id, 'players.name': {'$ne': player_name}, 'quests': []},
    {'$push': {'players': helpers.create_player(player_id, player_name)}}
  )
  return bool(result.modified_count)

def start_game(game_id, player_id, player_name, role_list, player_order):
  shuffled_roles = random.sample(role_list, len(role_list))
  players = [(helpers.URL_CHARS[i], name, shuffled_roles[i]) for i, name in enumerate(player_order)]
  _, leader, _ = random.choice(players)
  result = games.update_one(
    {
      'id': game_id,
      'creator': player_name,
      'players': {'$size': len(players), '$elemMatch': {'id': player_id, 'name': player_name}},
      'players.name': {'$all': [name for _, name, _ in players]},
      'quests': [],
    },
    {
      '$set': flatten_dicts([
        {'players.$[{}].role'.format(i): role, 'players.$[{}].order'.format(i): i} for i, _, role in players
      ]),
      '$push': {'quests': helpers.create_quest(1, 1, leader, len(players))}
    },
    array_filters = [{'{}.name'.format(i): name} for i, name, _ in players]
  )
  return bool(result.modified_count)

def propose_quest(quest_id, player_id, player_name, proposed_members):
  result = games.update_one(
    {
      'players.name': {'$all': proposed_members},
      'players': {'$elemMatch': {'id': player_id, 'name': player_name}},
      'quests': {'$elemMatch': {'id': quest_id, 'members': [], 'leader': player_name, 'size': len(proposed_members)}}
    },
    {'$set': {'quests.$[quest].members': proposed_members}},
    array_filters = [{'quest.id': quest_id}]
  )
  return bool(result.modified_count)

def vote_for_proposal(quest_id, player_id, player_name, vote):
  game = games.find_one_and_update(
    {
      'players': {'$elemMatch': {'id': player_id, 'name': player_name}},
      'quests': {'$elemMatch': {'id': quest_id, 'members': {'$ne': []}, 'votes.player': {'$ne': player_name}}}
    },
    {
      '$push': {'quests.$[quest].votes': {'player': player_name, 'vote': vote}},
      '$inc': {'quests.$[quest].remainingVotes': -1, 'quests.$[quest].voteStatus': vote}
    },
    array_filters = [{'quest.id': quest_id}],
    return_document = pymongo.collection.ReturnDocument.AFTER
  )
  if not game:
    return False

  # move to next leader if proposal was rejected
  quest = next(quest for quest in game['quests'] if quest['id'] == quest_id)
  next_quest = helpers.create_quest(
    quest['roundNumber'],
    quest['attemptNumber'] + 1,
    helpers.get_next_leader(game, quest),
    len(game['players'])
  )
  games.update_one(
    {'quests': {'$elemMatch': {'id': quest_id, 'remainingVotes': 0, 'voteStatus': {'$lte': 0}, 'attemptNumber': {'$lt': 5}}}},
    {'$push': {'quests': next_quest}}
  )
  return True

def vote_in_quest(quest_id, player_id, player_name, vote):
  game = games.find_one_and_update(
    {
      'players': {'$elemMatch': {'id': player_id, 'name': player_name, 'role': {'$in': helpers.roles_for_vote(vote)}}},
      'quests': {'$elemMatch': {'id': quest_id, 'members': player_name, 'results.player': {'$ne': player_name}, 'remainingVotes': 0, 'voteStatus': {'$gt': 0}}}
    },
    {
      '$push': {'quests.$[quest].results': {'player': player_name, 'vote': vote}},
      '$inc': {'quests.$[quest].remainingResults': -1, 'quests.$[quest].failures': 1 if vote < 0 else 0}
    },
    array_filters = [{'quest.id': quest_id}],
    return_document = pymongo.collection.ReturnDocument.AFTER
  )
  if not game:
    return False

  # move on to next round if all results have been collected (or do nothing if game is over)
  quest = next(quest for quest in game['quests'] if quest['id'] == quest_id)
  next_quest = helpers.create_quest(
    quest['roundNumber'] + 1,
    1,
    helpers.get_next_leader(game, quest),
    len(game['players'])
  )
  games.update_one(
    {'quests': {'$elemMatch': {'id': quest_id, 'remainingResults': 0, 'roundNumber': {'$lt': 5}}}},
    {'$push': {'quests': next_quest}}
  )
  return True

def get_state(game_id, player_id):
  game = games_cache.get_doc(game_id)
  if game is None or not any(player['id'] == player_id for player in game['players']):
    return False
  player = next(player for player in game['players'] if player['id'] == player_id)
  return helpers.get_state(game, player)

def flatten_dicts(dicts):
  return {k: v for d in dicts for k, v in d.items()}

File: store.js
const utils = require('./utils')

const ALPHABET = 'abcdefghijklmnopqrstuvwxyz'

// which roles have knowledge of which other roles
const roleKnowledge = {
  'merlin': ['morgana', 'assassin', 'minion', 'oberon'],
  'percival': ['merlin', 'morgana'],
  'loyal servant': [],
  'morgana': ['mordred', 'assassin', 'minion'],
  'mordred': ['morgana', 'assassin', 'minion'],
  'oberon': [],
  'assassin': ['morgana', 'mordred', 'minion'],
  'minion': ['morgana', 'mordred', 'assassin', 'minion']
}

// which characters are evil
const evilRoles = ['morgana', 'mordred', 'oberon', 'assassin', 'minion']

// how many people go on quests per round based on the number of players
const questConfigurations = {
  5: [2, 3, 2, 3, 3],
  6: [2, 3, 4, 3, 4],
  7: [2, 3, 3, 4, 4],
  8: [3, 4, 4, 5, 5],
  9: [3, 4, 4, 5, 5],
  10: [3, 4, 4, 5, 5]
}

exports.GameModel = (games) => ({
  createGame: (gameId, playerId, playerName) => {
    return wrap(games.insertOne(
      {
        id: gameId,
        creator: playerName,
        players: [createPlayer(playerId, playerName)],
        playerOrder: [],
        questConfiguration: [],
        currentQuest: null,
        questHistory: [],
        assassinTarget: null
     }))
  },
  joinGame: (gameId, playerId, playerName) => {
    return wrap(games.updateOne(
      {id: gameId, 'players.name': {$ne: playerName}, currentQuest: null},
      {$push: {players: createPlayer(playerId, playerName)}}
    ))
  },
  startGame: (gameId, playerId, playerName, roleList, playerOrder) => {
    const shuffledRoles = utils.shuffle(roleList)
    const players = playerOrder.map((name, i) => ({order: ALPHABET[i], name, role: shuffledRoles[i]}))
    const leader = playerOrder[Math.floor(Math.random() * playerOrder.length)]
    const startingQuest = {
      id: utils.randomId(),
      roundNumber: 1,
      attemptNumber: 1,
      size: questConfigurations[players.length][0],
      leader: leader,
      members: [],
      votes: [],
      results: [],
      remainingVotes: players.length,
      voteStatus: 0,
      remainingResults: questConfigurations[players.length][0],
      failures: 0
    }

    return wrap(games.updateOne(
      {
        id: gameId,
        creator: playerName,
        players: {$size: players.length, $elemMatch: {id: playerId, name: playerName}},
        'players.name': {$all: playerOrder},
        currentQuest: null,
      },
      {
        $set: {
          currentQuest: startingQuest,
          playerOrder: utils.rotate(playerOrder, playerOrder.indexOf(leader)),
          questConfiguration: questConfigurations[players.length],
          ...utils.flatMap(players, ({order, role}) => (
            {[`players.$[${order}].role`]: role, [`players.$[${order}].order`]: order}
          ))
        }
      },
      {arrayFilters: players.map(({order, name}) => ({[`${order}.name`]: name}))}
    ))
  },
  proposeQuest: (questId, playerId, playerName, proposedMembers) => {
    return wrap(games.updateOne(
      {
        'players.name': {$all: proposedMembers},
        players: {$elemMatch: {id: playerId, name: playerName}},
        'currentQuest.id': questId,
        'currentQuest.members': [],
        'currentQuest.leader': playerName,
        'currentQuest.size': proposedMembers.length
      },
      {$set: {'currentQuest.members': proposedMembers}}
    ))
  },
  voteForProposal: (questId, playerId, playerName, vote) => {
    const proposalRejected = {
      $and: [
        {$eq: ['$currentQuest.remainingVotes', 0]},
        {$lte: ['$currentQuest.voteStatus', 0]},
        {$lt: ['$currentQuest.attemptNumber', 5]}
      ]
    }
    const nextQuestAttempt = {
      id: utils.randomId(),
      roundNumber: '$currentQuest.roundNumber',
      attemptNumber: {$add: ['$currentQuest.attemptNumber', 1]},
      size: '$currentQuest.size',
      leader: {$arrayElemAt: ['$playerOrder', {$mod: [{$add: [{$size: '$questHistory'}, 1]}, {$size: '$players'}]}]},
      members: [],
      votes: [],
      results: [],
      remainingVotes: {$size: '$players'},
      voteStatus: 0,
      remainingResults: '$currentQuest.size',
      failures: 0
    }
    return wrap(games.updateOne(
      {
        players: {$elemMatch: {id: playerId, name: playerName}},
        'currentQuest.id': questId,
        'currentQuest.members': {$ne: []},
        'currentQuest.votes.player': {$ne: playerName}
      },
      [
        {
          $addFields: {
            'currentQuest.votes': {$concatArrays: ['$currentQuest.votes', [{player: playerName, vote: vote}]]},
            'currentQuest.remainingVotes': {$add: ['$currentQuest.remainingVotes', -1]},
            'currentQuest.voteStatus': {$add: ['$currentQuest.voteStatus', vote]}
          }
        },
        {
          $addFields: {
            currentQuest: {$cond: [proposalRejected, nextQuestAttempt, '$currentQuest']},
            questHistory: {$cond: [proposalRejected, {$concatArrays: ['$questHistory', ['$currentQuest']]}, '$questHistory']}
          }
        }
      ]
    ))
  },
  voteInQuest: (questId, playerId, playerName, vote) => {
    const votingComplete = {
      $and: [
        {$eq: ['$currentQuest.remainingResults', 0]},
        {$lt: ['$currentQuest.roundNumber', 5]}
      ]
    }
    const nextQuestRound = {
      id: utils.randomId(),
      roundNumber: {$add: ['$currentQuest.roundNumber', 1]},
      attemptNumber: 1,
      size: {$arrayElemAt: ['$questConfiguration', '$currentQuest.roundNumber']},
      leader: {$arrayElemAt: ['$playerOrder', {$mod: [{$add: [{$size: '$questHistory'}, 1]}, {$size: '$players'}]}]},
      members: [],
      votes: [],
      results: [],
      remainingVotes: {$size: '$players'},
      voteStatus: 0,
      remainingResults: {$arrayElemAt: ['$questConfiguration', '$currentQuest.roundNumber']},
      failures: 0
    }
    return wrap(games.updateOne(
      {
        players: {$elemMatch: {id: playerId, name: playerName, role: {$in: rolesForVote(vote)}}},
        'currentQuest.id': questId,
        'currentQuest.members': playerName,
        'currentQuest.results.player': {$ne: playerName},
        'currentQuest.remainingVotes': 0,
        'currentQuest.voteStatus': {$gt: 0}
      },
      [
        {
          $addFields: {
            'currentQuest.results': {$concatArrays: ['$currentQuest.results', [{player: playerName, vote: vote}]]},
            'currentQuest.remainingResults': {$add: ['$currentQuest.remainingResults', -1]},
            'currentQuest.failures': {$add: ['$currentQuest.failures', vote < 0 ? 1 : 0]}
          }
        },
        {
          $addFields: {
            currentQuest: {$cond: [votingComplete, nextQuestRound, '$currentQuest']},
            questHistory: {$cond: [votingComplete, {$concatArrays: ['$questHistory', ['$currentQuest']]}, '$questHistory']}
          }
        }
      ]
    ))
  },
  assassinate: (gameId, playerId, playerName, target) => {
    return wrap(games.updateOne(
      {
        id: gameId,
        players: {$elemMatch: {id: playerId, name: playerName, role: 'assassin'}},
        assassinTarget: null
      },
      {$set: {assassinTarget: target}}
    ))
  },
  fetchState: (playerId) => {
    return games.findOne({'players.id': playerId}).then(game => {
      if (!game) {
        return {error: 'Invalid operation'}
      }
      return getState(game, utils.find(game.players, player => player.id === playerId))
    })
  },
  onUpdate: (fn) => {
    games.watch({fullDocument: 'updateLookup'}).on('change', data => {
      const game = data.fullDocument
      game.players.forEach(player => fn(player.id, getState(game, player)))
    })
  }
})

const wrap = (promise) => promise.then(result => {
  if (result.modifiedCount === 0 || result.insertedCount === 0) {
    return {error: 'Invalid operation'}
  }
  return {error: null}
})

const createPlayer = (playerId, playerName) => ({
  id: playerId, name: playerName, role: null, order: 0
})

const rolesForVote = (vote) => vote < 0 ? evilRoles : Object.keys(roleKnowledge)

const getState = (game, player) => {
  const players = utils.sortBy(game.players, 'order')
  return {
    'id': game.id,
    'creator': game.creator,
    'players': players.map(p => p.name),
    'roles': game.currentQuest ? players.map(p => p.role).sort() : [],
    'questConfigurations': game.questConfiguration,
    'myId': player.id,
    'myName': player.name,
    'myRole': player.role,
    'knowledge': getPlayerKnowledge(game, player),
    'currentQuest': game.currentQuest ? sanitizeQuest(game, game.currentQuest, player) : null,
    'questHistory': game.questHistory.map(quest => sanitizeQuest(game, quest, player)),
    'status': getGameStatus(game),
    'allRoles': Object.keys(roleKnowledge).map(role => ({role, evil: evilRoles.includes(role)}))
  }
}

const getPlayerKnowledge = (game, player) => {
  const knowledge = roleKnowledge[player.role] || []
  const knownPlayers = game.players.filter(p => p.id !== player.id && knowledge.includes(p.role))
  return {
    players: knownPlayers.map(p => p.name).sort(),
    roles: knownPlayers.map(p => p.role).sort()
  }
}

const sanitizeQuest = (game, quest, player) => {
  const q = Object.assign({}, quest)
  q.votes = quest.remainingVotes === 0 ? utils.sortBy(quest.votes, 'player') : []
  q.results = quest.remainingResults === 0 ? quest.results.map(result => result.vote).sort() : []
  q.myVote =  utils.find(quest.votes, vote => vote.player === player.name).vote
  q.myResult = utils.find(quest.results, result => result.player === player.name).vote
  q.status = getQuestStatus(game, quest)
  delete q.voteStatus
  delete q.failures
  return q
}

const getGameStatus = (game) => {
  if (!game.currentQuest)
    return 'not_started'
  if (game.currentQuest.attemptNumber === 5 && game.currentQuest.remainingVotes === 0)
    return 'evil_won'
  if (game.questHistory.filter(quest => getQuestStatus(game, quest) === 'failed').length > 2)
    return 'evil_won'
  if (game.questHistory.filter(quest => getQuestStatus(game, quest) === 'passed').length <= 2)
    return 'in_progress'
  if (game.players.some(player => player.role === 'assassin') && game.assassinTarget === null)
    return 'assassinating'
  if (game.assassinTarget === utils.find(game.players, player => player.role === 'merlin').name)
    return 'evil_won'
  return 'good_won'
}

const getQuestStatus = (game, quest) => {
  if (quest.members.length === 0)
    return 'proposing_quest'
  if (quest.remainingVotes > 0)
    return 'voting_for_proposal'
  if (quest.voteStatus <= 0)
    return 'proposal_rejected'
  if (quest.remainingResults > 0)
    return 'voting_in_quest'
  return didQuestPass(game, quest) ? 'passed' : 'failed'
}

const didQuestPass = (game, quest) => {
  return quest.failures === 0 || (quest.roundNumber === 4 && game.players.length > 6 && quest.failures === 1)
}
